tokio-compat/src/runtime/threadpool/mod.rs

```diff
@@ -0,0 +1,405 @@
+}
+
+#[derive(Clone)]
+struct CompatSender<S>(S);
+
+// ===== impl Runtime =====
+
+/// Start the Tokio runtime using the supplied future to bootstrap execution.
+///
+/// This function is used to bootstrap the execution of a Tokio application. It
+/// does the following:
+///
+/// * Start the Tokio runtime using a default configuration.
+/// * Spawn the given future onto the thread pool.
+/// * Block the current thread until the runtime shuts down.
+///
+/// Note that the function will not return immediately once `future` has
+/// completed. Instead it waits for the entire runtime to become idle.
+///
+/// See the [module level][mod] documentation for more details.
+///
+/// # Examples
+///
+/// ```rust
+/// use futures_01::{Future as Future01, Stream as Stream01};
+/// use tokio_01::net::TcpListener;
+///
+/// # fn process<T>(_: T) -> Box<Future01<Item = (), Error = ()> + Send> {
+/// # unimplemented!();
+/// # }
+/// # fn dox() {
+/// # let addr = "127.0.0.1:8080".parse().unwrap();
+/// let listener = TcpListener::bind(&addr).unwrap();
+///
+/// let server = listener.incoming()
+///     .map_err(|e| println!("error = {:?}", e))
+///     .for_each(|socket| {
+///         tokio_01::spawn(process(socket))
+///     });
+///
+/// tokio_compat::run(server);
+/// # }
+/// # pub fn main() {}
+/// ```
+///
+/// # Panics
+///
+/// This function panics if called from the context of an executor.
+///
+/// [mod]: ../index.html
+pub fn run<F>(future: F)
+where
+    F: Future01<Item = (), Error = ()> + Send + 'static,
+{
+    run_03(future.compat().map(|_| ()))
+}
+
+/// Start the Tokio runtime using the supplied `std::future` future to bootstrap execution.
+///
+/// This function is used to bootstrap the execution of a Tokio application. It
+/// does the following:
+///
+/// * Start the Tokio runtime using a default configuration.
+/// * Spawn the given future onto the thread pool.
+/// * Block the current thread until the runtime shuts down.
+///
+/// Note that the function will not return immediately once `future` has
+/// completed. Instead it waits for the entire runtime to become idle.
+///
+/// See the [module level][mod] documentation for more details.
+///
+/// # Examples
+///
+/// ```rust
+/// use futures_01::{Future as Future01, Stream as Stream01};
+/// use tokio_01::net::TcpListener;
+///
+/// # fn process<T>(_: T) -> Box<Future01<Item = (), Error = ()> + Send> {
+/// # unimplemented!();
+/// # }
+/// # fn dox() {
+/// # let addr = "127.0.0.1:8080".parse().unwrap();
+/// let listener = TcpListener::bind(&addr).unwrap();
+///
+/// let server = listener.incoming()
+///     .map_err(|e| println!("error = {:?}", e))
+///     .for_each(|socket| {
+///         tokio_01::spawn(process(socket))
+///     });
+///
+/// tokio_compat::run(server);
+/// # }
+/// # pub fn main() {}
+/// ```
+///
+/// # Panics
+///
+/// This function panics if called from the context of an executor.
+///
+/// [mod]: ../index.html
+pub fn run_03<F>(future: F)
```

Comment:
I thought it was best for the `run` function to take a `futures` 0.1 `Future`, so that the compat runtime can be drop-in compatible with code using the `tokio` 0.1 runtime. However, we also need a `run` for futures 0.3 futures, as an entry-point for new code that needs to call into legacy libraries (especially since `tokio::main` doesn't support the compat runtime yet).

I also considered calling this `run_async`, but I thought `run_03` was more accurate. Open to considering other names though!

Reply:
I actually think maybe async is better here since 0.3 loses its point since the future trait is in the standard library?

Reply:
I like either `run_async`, or `run_std_future`. Something to be clear that these are the `async`/`await` futures from standard Rust, not from a futures crate